File: examples/android/MSDKSampleAppJava/app/src/main/java/msdk/java/sample/MainActivity.java
package msdk.java.sample;

import android.os.Bundle;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import org.json.JSONException;

import java.util.concurrent.Executors;

import msdk.java.sample.databinding.MainActivityBinding;
import msdk.java.handlers.Initialization;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        MainActivityBinding viewBinding = MainActivityBinding.inflate(getLayoutInflater());
        setContentView(viewBinding.getRoot());

        MainPagerAdapter adapter = new MainPagerAdapter(getSupportFragmentManager());
        viewBinding.pager.setAdapter(adapter);
        viewBinding.tabLayout.setupWithViewPager(viewBinding.pager);

        initialize();
    }

    private void initialize() {
        Toast.makeText(this, "Application initialization started", Toast.LENGTH_SHORT).show();
        Executors.newSingleThreadExecutor().execute(() -> {
            // 1. Start initialization
            Initialization.Constants constants = buildConstants();
            Initialization.initialize(this, constants, R.raw.genesis).handleAsync((res, err) -> {
                if (err != null) {
                    runOnUiThread(() -> Toast.makeText(this, "Application was not initialized!", Toast.LENGTH_SHORT).show());
                } else {
                    runOnUiThread(() -> Toast.makeText(this, "Application successfully initialized", Toast.LENGTH_SHORT).show());
                }
                return null;
            });
        });
    }

    private Initialization.Constants buildConstants() {
        return Initialization.Constants.builder()
                .withAgencyEndpoint(Constants.AGENCY_ENDPOINT)
                .withAgencyDid(Constants.AGENCY_DID)
                .withAgencyVerkey(Constants.AGENCY_VERKEY)
                .withWalletName(Constants.WALLET_NAME)
                .withPrefsName(Constants.PREFERENCES_KEY)
                .withSponseeId(Constants.SPONSEE_ID_KEY)
                .withServerUrl(Constants.SERVER_URL)
                .withLogo(Constants.LOGO)
                .withName(Constants.NAME)
                .build();
    }
}

File: examples/android/MSDKSampleAppKotlin/app/src/main/java/msdk/kotlin/sample/MainActivity.kt
package msdk.kotlin.sample

import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import msdk.kotlin.sample.databinding.MainActivityBinding
import msdk.kotlin.sample.handlers.Initialization
import java.util.concurrent.Executors


class MainActivity : AppCompatActivity() {
    private lateinit var viewBinding: MainActivityBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initViews()
        initialize()
    }

    private fun initViews() {
        viewBinding = MainActivityBinding.inflate(layoutInflater)
        setContentView(viewBinding.root)
        val adapter = MainPagerAdapter(supportFragmentManager)
        viewBinding.pager.adapter = adapter
        viewBinding.tabLayout.setupWithViewPager(viewBinding.pager)
    }

    private fun initialize() {
        Toast.makeText(this, "Application initialization started", Toast.LENGTH_SHORT).show()
        Executors.newSingleThreadExecutor().execute {
            // 1. Start initialization
            val constants = buildConstants()
            Initialization.initialize(this, constants, R.raw.genesis)
                .whenComplete { res: Void?, ex: Throwable? ->
                    if (ex != null) {
                        runOnUiThread {
                            Toast.makeText(this, "Application initialization failed", Toast.LENGTH_SHORT).show()
                        }
                    } else {
                        runOnUiThread {
                            Toast.makeText(this, "Application successfully initialized", Toast.LENGTH_SHORT).show()
                        }
                    }
                }
        }
    }

    private fun buildConstants(): Initialization.Constants {
        return Initialization.Constants.builder()
                .withAgencyEndpoint(Constants.AGENCY_ENDPOINT)
                .withAgencyDid(Constants.AGENCY_DID)
                .withAgencyVerkey(Constants.AGENCY_VERKEY)
                .withWalletName(Constants.WALLET_NAME)
                .withPrefsName(Constants.PREFS_NAME)
                .withSponseeId(Constants.SPONSEE_ID)
                .withServerUrl(Constants.SERVER_URL)
                .withLogo(Constants.LOGO)
                .withName(Constants.NAME)
                .build()
    }
}
